### PR TITLE
Add implicit x flag when morphing copy block

### LINKED ISSUE #13882: VectorExp_ro test fails in Debug config when JIT is in minopts mode
I would expect our JIT tests to either:
1) Run correctly when COMPLUS_JitMinOpts=1 has been specified
2) Specify JitOptimizationSensitive in their project file to indicate they are unsafe to run with non-default optimizations.

VectorExp_ro doesn't appear to do either of those things. The failure looks like this:

```
F:\github\coreclr\tests>set CORE_ROOT=F:\github\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root
F:\github\coreclr\tests>set COMPLUS_JitMinOpts=1
F:\github\coreclr\tests>F:\github\coreclr\bin\tests\Windows_NT.x64.Debug\JIT\SIMD\VectorExp_ro\VectorExp_ro.cmd
BEGIN EXECUTION
 "F:\github\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe" VectorExp_ro.exe
Missing flags on tree [000208]: --X--------
               [000208] -A--G+------             *  IND       simd32
               [000207] ----G+------             |     /--*  ADDR      byref
               [000054] R---G+------             |     |  \--*  OBJ(32)   simd32
               [000053] ----G+------             |     |     \--*  ADDR      byref
               [000041] ----G+------             |     |        |     /--*  LCL_VAR   simd32(AX) V06 loc2
               [000044] ----G+------             |     |        |  /--*  SIMD      simd32 long &~
               [000043] -----+------             |     |        |  |  \--*  LCL_VAR   simd32 V09 tmp1
               [000045] ----G+-N----             |     |        \--*  SIMD      simd32 long |
               [000037] -----+------             |     |           |  /--*  LCL_VAR   simd32 V09 tmp1
               [000042] ----G+------             |     |           \--*  SIMD      simd32 long &
               [000039] ----G+------             |     |              |  /--*  LCL_VAR   simd32(AX) V06 loc2
               [000040] ----G+------             |     |              \--*  SIMD      simd32 long -
               [000038] -----+------             |     |                 \--*  LCL_VAR   simd32 V08 tmp0
               [000046] -A--G+------             |  /--*  COMMA     byref
               [000026] ----G+------             |  |  |     /--*  LCL_VAR   simd32(AX) V06 loc2
               [000033] ----G+-N----             |  |  |  /--*  SIMD      simd32 long gt
               [000032] -----+------             |  |  |  |  \--*  LCL_VAR   simd32 V08 tmp0
               [000036] -A--G+--R---             |  |  \--*  ASG       simd32 (copy)
               [000034] D----+-N----             |  |     \--*  LCL_VAR   simd32 V09 tmp1
               [000047] -A--G+-N----             \--*  COMMA     byref
               [000028] -----+-N----                |  /--*  SIMD      simd32 long init
               [000027] -----+------                |  |  \--*  CNS_INT   long   0
               [000031] -A---+--R---                \--*  ASG       simd32 (copy)
               [000029] D----+-N----                   \--*  LCL_VAR   simd32 V08 tmp0

Assert failure(PID 24540 [0x00005fdc], Thread: 12360 [0x3048]): Assertion failed '!"Missing flags on tree"' in 'VectorExpTest`1[Int64][System.Int64]:VectorExp(struct,long,long,long):int' (IL size 148)

    File: f:\github\coreclr\src\jit\flowgraph.cpp Line: 21361
    Image: F:\github\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\CoreRun.exe

Expected: 100
Actual: 123456789
END EXECUTION - FAILED
FAILED
```

### PR DESCRIPTION
**Pull request by jashook**
Fixes #13882 

/cc @sandreenko

### CODE REVIEW

**Comment by mmitche** [comment]
@dotnet-bot test this please

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -10772,6 +10778,9 @@ GenTreePtr Compiler::fgMorphCopyBlock(GenTreePtr tree)
             rhs             = fgMorphBlockOperand(rhs, asgType, blockWidth, false /*!isDest*/);
             asg->gtOp.gtOp2 = rhs;
 
+            // Propogate flags up
+            asg->gtFlags |= (rhs->gtFlags & GTF_ALL_EFFECT);
```
Comment (sandreenko) [inline src/jit/morph.cpp]: There is another place in `fgMorphCopyBlock` where we assign `asg->gtOp.gtOp2 = rhs;`
Do we need to updates flags there as well?

Also, it looks like we protect against 'missed flag on a tree', but we do not protect against extra flag, if some side effects were deleted from children. However, it is not important for now.

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -9941,6 +9941,12 @@ GenTreePtr Compiler::fgMorphBlkToInd(GenTreeBlk* tree, var_types type)
 {
     tree->SetOper(GT_IND);
     tree->gtType = type;
+
+    if (tree->OperMayThrow(this))
```
[the remark this answers is not in the document]
Comment (briansull) [reply src/jit/morph.cpp]: We started with a GT_BLK node that never throws an exception, right?

So changing this  should result in an GT_IND that also never throw an exception.

[the remark this answers is not in the document]
Comment (jashook) [reply src/jit/morph.cpp]: Then the correct fix is to conditionally ignore the assert or to remove the check if ... OperMayThrow while checking flags.

Opinions?

File: src/jit/morph.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/jit/morph.cpp]: > we do not protect against extra flag, if some side effects were deleted from children. However, it is not important for now.

I believe that this is a known issue, i.e. we don't re-analyze the entire tree when a side effect is deleted.

File: src/jit/morph.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/jit/morph.cpp]: Morphing a `GT_BLK` to a `GT_IND` should not change the value of the `GTF_EXCEPT` flag. Where is the assert happening? Is it a case where it should have been set on the original `GT_BLK`, or a case where some assert is too conservative?

[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/jit/morph.cpp]: From the example in #13882, it seems that, (perhaps due to some morphing or opts not happening?) we miss the fact that the "indirection" is of a SIMD result, which will never throw.

**Comment by sandreenko** [comment]
> I believe that this is a known issue, i.e. we don't re-analyze the entire tree when a side effect is deleted.

What do you mean? We do not need to re-analyze the entire tree and we care about correct flags. Some recent fixes about that: #14179, #14102, #13863, #13838 .

**Comment by CarolEidt** [comment]
> The problem is that exception flag is not set on block node.

But it appears from the listing that the block node is a BLK(ADDR(SIMD)) which doesn't throw, so the exception flag should not be set.

**Comment by CarolEidt** [comment]
> What do you mean? We do not need to re-analyze the entire tree and we care about correct flags.

Yes, we care about the flags being correct. However, I don't believe it's the case that we ensure that in *all* cases a flag removal is propagated up as far as it needs to be (e.g. through commas, etc.) It may be that those issues have all been addressed, but I didn't think it was a guarantee.

**Comment by sandreenko** [comment]
> > The problem is that exception flag is not set on block node.
> 
> But it appears from the listing that the block node is a BLK(ADDR(SIMD)) which doesn't throw, so the exception flag should not be set.

If you ask node 208 before `fgMorphBlkToInd ` `208->OperMayThrow(this);` it will return yes. It means flag is missed there.

File: src/jit/morph.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (briansull) [reply src/jit/morph.cpp]: Or arrange for GTF_IND_NONFAULTING to be set.

**Comment by CarolEidt** [comment]
> If you ask node 208 before fgMorphBlkToInd 208->OperMayThrow(this); it will return yes. It means flag is missed there.

But my point is that it should not return yes, if the only nodes under it are SIMD nodes. Those are never actually indirections, so they won't throw.

**Comment by sandreenko** [comment]
I am leaving now, but I suspect that this dump already shows the error:

```
In Compiler::impImportCall: opcode is call, kind=0, callRetType is bool, structSize is 0
  Known type Vector
SIMD Candidate Type System.Numerics.Vector
 Found type Vector
  Known type SIMD Vector<Long>
Method LessThanOrEqualAll is NOT a SIMD intrinsic
  Known type SIMD Vector<Long>
  Known type SIMD Vector<Long>

    [ 1]  34 (0x022) brtrue.s

               [000059] ------------             *  STMT      void  (IL 0x016...  ???)
               [000058] -ACXG-------             \--*  JTRUE     void
               [000056] ------------                |  /--*  CNS_INT   int    0
               [000057] -ACXG-------                \--*  NE        int
               [000049] -ACXG-------                   \--*  CALL      int    System.Numerics.Vector.LessThanOrEqualAll
               [000054] ---XG-------                      |     /--*  OBJ(32)   simd32
               [000053] ------------                      |     |  \--*  ADDR      byref
               [000041] ------------                      |     |     |     /--*  LCL_VAR   simd32 V06 loc2
               [000044] ------------                      |     |     |  /--*  SIMD      simd32 long &~
               [000043] ------------                      |     |     |  |  \--*  LCL_VAR   simd32 V09 tmp1
               [000045] ------------                      |     |     \--*  SIMD      simd32 long |
               [000037] ------------                      |     |        |  /--*  LCL_VAR   simd32 V09 tmp1
               [000042] ------------                      |     |        \--*  SIMD      simd32 long &
               [000039] ------------                      |     |           |  /--*  LCL_VAR   simd32 V06 loc2
               [000040] ------------                      |     |           \--*  SIMD      simd32 long -
               [000038] ------------                      |     |              \--*  LCL_VAR   simd32 V08 tmp0
               [000046] -A----------                      |  /--*  COMMA     simd32
               [000026] ------------                      |  |  |     /--*  LCL_VAR   simd32 V06 loc2
               [000033] -------N----                      |  |  |  /--*  SIMD      simd32 long gt
               [000032] ------------                      |  |  |  |  \--*  LCL_VAR   simd32 V08 tmp0
               [000036] -A------R---                      |  |  \--*  ASG       simd32 (copy)
               [000034] D------N----                      |  |     \--*  LCL_VAR   simd32 V09 tmp1
               [000047] -A-X-------- arg0                 +--*  COMMA     simd32
               [000028] -------N----                      |  |  /--*  SIMD      simd32 long init
               [000027] ------------                      |  |  |  \--*  CNS_INT   long   0
               [000031] -A------R---                      |  \--*  ASG       simd32 (copy)
               [000029] D------N----                      |     \--*  LCL_VAR   simd32 V08 tmp0
               [000051] R----------- arg1                 \--*  OBJ(32)   simd32
               [000050] L-----------                         \--*  ADDR      byref
               [000048] ------------                            \--*  LCL_VAR   simd32 V07 loc3
```

If I understand our flag system 000046 must be marked with X in this case, because it is child 000054 is marked with X.

It happens in `impImportCall` that I was not allowed to clean.

**Comment by jashook** [comment]
I will close the for now, I follow Carol and Brian’s logic and believe the flag should remain under. The check for flags assumes incorrectly the flag should be set.

Thanks!

**Pull request closed without merging**